### PR TITLE
feature: hardening type-safety

### PR DESCRIPTION
**Pull request by boesing**
|    Q          |   A
|-------------- | ------
| BC Break      | yes
| New Feature   | kinda
| QA            | yes

### Description

- Removed psalm baseline
- Added psalm annotations to ensure types
- Supersedes #13

### CODE REVIEW

**Comment by boesing** [comment]
Codesniffer issues are fixed with #14

File: src/Emitter/EmitterStack.php [hunk 1]
```diff
@@ -45,56 +56,29 @@ public function emit(ResponseInterface $response) : bool
...
-    /**
-     * Set an emitter on the stack by index.
-     *
-     * @param mixed $index
-     * @param EmitterInterface $emitter
-     * @return void
-     * @throws InvalidArgumentException if not an EmitterInterface instance
-     */
-    public function offsetSet($index, $emitter)
+    public function offsetSet(int $index, EmitterInterface $emitter): void
```
Comment (lcobucci) [inline src/Emitter/EmitterStack.php]: `SplStack` implements a few interfaces that use this and other methods to integrate more easily to the language. This change is then a BC-break: https://3v4l.org/rNtHA#v8.0.8 vs https://3v4l.org/ZgTsb#v8.0.8

To solve that, we'd need to implement `Iterator`, `ArrayAccess`, and `Countable`: https://3v4l.org/CuTCA#v8.0.8

This also means we must keep the parameters covariant (`ArrayAccess` defines `$offset` and `$value` as `mixed`).

File: src/Emitter/SapiEmitterTrait.php [hunk 2]
```diff
@@ -79,8 +79,9 @@ private function emitHeaders(ResponseInterface $response): void
         $statusCode = $response->getStatusCode();
 
         foreach ($response->getHeaders() as $header => $values) {
+            assert(is_string($header));
             $name  = $this->filterHeader($header);
-            $first = $name === 'Set-Cookie' ? false : true;
+            $first = ! ($name === 'Set-Cookie');
```
Comment (lcobucci) [inline src/Emitter/SapiEmitterTrait.php]: I think this is easier to read:

```suggestion
            $first = $name !== 'Set-Cookie';
```

File: src/Emitter/SapiStreamEmitter.php [hunk 3]
```diff
@@ -78,10 +81,11 @@ private function emitBody(ResponseInterface $response) : void
 
     /**
      * Emit a range of the message body.
+     * @psalm-param ParsedRangeType $range
      */
     private function emitBodyRange(array $range, ResponseInterface $response) : void
     {
-        list($unit, $first, $last, $length) = $range;
+        list($_, $first, $last, $_) = $range;
```
Comment (lcobucci) [inline src/Emitter/SapiStreamEmitter.php]: I don't think we need `$_` here.

```suggestion
        [, $first, $last] = $range;
```

File: src/Emitter/EmitterStack.php (hunk 1 above)
Comment (boesing) [reply src/Emitter/EmitterStack.php]: I've rollbacked this and added proper annotations instead.

File: src/Emitter/SapiEmitterTrait.php (hunk 2 above)
Comment (boesing) [reply src/Emitter/SapiEmitterTrait.php]: Yeah, phpstorm suggestion 😅

File: src/Emitter/SapiStreamEmitter.php (hunk 3 above)
Comment (boesing) [reply src/Emitter/SapiStreamEmitter.php]: Oh yah, haven't really seen that it is `list`. Just wanted to fix the unused variable issue 😂

File: src/Emitter/EmitterStack.php (hunk 1 above)
Comment (boesing) [reply src/Emitter/EmitterStack.php]: BTW: targeting 2.0 as a new major and thus BC breaks should be fine. But I guess that we can also live with some psalm suppressions 👍🏼

File: test/Emitter/SapiStreamEmitterTest.php [hunk 4]
```diff
@@ -53,7 +56,7 @@ public function testEmitCallbackStreamResponse(): void
             ->withBody($stream);
         ob_start();
         $this->emitter->emit($response);
-        $this->assertSame('it works', ob_get_clean());
+        self::assertSame('it works', ob_get_clean());
```
Comment (lcobucci) [inline test/Emitter/SapiStreamEmitterTest.php:59]: Oh, I hadn't seen this one 😁

PHPUnit has an assertion to verify the output of the execution. It's something that we should address in another PR, though.

File: psalm.xml.dist [hunk 5]
```diff
@@ -1,10 +1,9 @@
 <?xml version="1.0"?>
 <psalm
-        totallyTyped="true"
+        errorLevel="1"
```
Comment (lcobucci) [inline psalm.xml.dist:3]: I'd suggest to keep the previous configuration so that new code follows a more strict setup.

**Review by lcobucci — APPROVED** [review]
I believe that a stricter config with baselined issues leads to better results but that doesn't necessarily block this PR IMHO.

File: psalm.xml.dist (hunk 5 above)
Comment (boesing) [reply psalm.xml.dist:3]: This is the most strictest level. 
totallyTyped option is deprecated.

https://psalm.dev/docs/running_psalm/configuration/#totallytyped

**Comment by boesing** [comment]
~I have to verify why unit tests now fail. They supposed to work.~

Was related to the refactoring regarding `list`, fixed now ✅

File: psalm.xml.dist (hunk 5 above)
Comment (lcobucci) [reply psalm.xml.dist:3]: TIL!

**Pull request merged**